In [ ]:
import os
from dotenv import load_dotenv

load_dotenv()

os.environ["OPENAI_API_KEY"] = os.getenv("OPENAI_API_KEY")
os.environ["ANTHROPIC_API_KEY"] = os.getenv("ANTHROPIC_API_KEY")
os.environ["GROQ_API_KEY"] = os.getenv("GROQ_API_KEY")

In [2]:
from autogen_agentchat.agents import AssistantAgent
from autogen_ext.models.openai import OpenAIChatCompletionClient

In [3]:
model_client = OpenAIChatCompletionClient(
    model='gpt-4o-mini',
    temperature=0.3
)

In [4]:
dsa_solver = AssistantAgent(
    name = 'Complex_DSA_Solver',
    model_client=model_client,
    description='A DSA solver',
    system_message="You give code in python to solve complex DSA problems. Give under 100 words."
)

code_reviewer = AssistantAgent(
    name = 'CODE_REVIEWER',
    model_client=model_client,
    description='A Code Reviewer',
    system_message="You review the code given by the complex_dsa_solver and make sure it is optimized.Give under 10 words. If you feel that the code is fine, please say 'TERMINATE'"
)

code_editor = AssistantAgent(
    name = 'CODE_EDITOR',
    model_client=model_client,
    description='A Code editor',
    system_message="You make the code easy to understand and add comments wherever required.Give under 10 words"
)

In [5]:
from autogen_agentchat.teams import RoundRobinGroupChat
from autogen_agentchat.messages import TextMessage

team = RoundRobinGroupChat(
    participants=[dsa_solver, code_reviewer, code_editor],
    max_turns=3   
)

team

In [ ]:
async def run_team():
    task = TextMessage(content='write a code to search an element in a sorted array.', source='user')

    result = await team.run(task=task)


    print(result)

await run_team()

messages=[TextMessage(id='91c37ddc-dd44-468a-a60a-c9b408a54f33', source='user', models_usage=None, metadata={}, created_at=datetime.datetime(2025, 7, 20, 18, 0, 17, 842190, tzinfo=datetime.timezone.utc), content='write a code to search an element in a sorted array.', type='TextMessage'), TextMessage(id='a2f6c64e-2cfc-46c5-b736-ceec359833db', source='Complex_DSA_Solver', models_usage=RequestUsage(prompt_tokens=42, completion_tokens=146), metadata={}, created_at=datetime.datetime(2025, 7, 20, 18, 0, 20, 420232, tzinfo=datetime.timezone.utc), content="Here's a Python implementation of binary search to find an element in a sorted array:\n\n```python\ndef binary_search(arr, target):\n    left, right = 0, len(arr) - 1\n    while left <= right:\n        mid = left + (right - left) // 2\n        if arr[mid] == target:\n            return mid\n        elif arr[mid] < target:\n            left = mid + 1\n        else:\n            right = mid - 1\n    return -1\n\n# Example usage\narr = [1, 2, 3

messages=[
    
    TextMessage(id='91c37ddc-dd44-468a-a60a-c9b408a54f33', source='user', models_usage=None, metadata={}, created_at=datetime.datetime(2025, 7, 20, 18, 0, 17, 842190, tzinfo=datetime.timezone.utc), content='write a code to search an element in a sorted array.', type='TextMessage'), 
    
    TextMessage(id='a2f6c64e-2cfc-46c5-b736-ceec359833db', source='Complex_DSA_Solver', models_usage=RequestUsage(prompt_tokens=42, completion_tokens=146), metadata={}, created_at=datetime.datetime(2025, 7, 20, 18, 0, 20, 420232, tzinfo=datetime.timezone.utc), content="Here's a Python implementation of binary search to find an element in a sorted array:\n\n```python\ndef binary_search(arr, target):\n    left, right = 0, len(arr) - 1\n    while left <= right:\n        mid = left + (right - left) // 2\n        if arr[mid] == target:\n            return mid\n        elif arr[mid] < target:\n            left = mid + 1\n        else:\n            right = mid - 1\n    return -1\n\n# Example usage\narr = [1, 2, 3, 4, 5]\ntarget = 3\nprint(binary_search(arr, target))  # Output: 2\n```", type='TextMessage'), 
    
    TextMessage(id='2118bdf0-a082-4831-97f8-abecce912d3c', source='CODE_REVIEWER', models_usage=RequestUsage(prompt_tokens=220, completion_tokens=3), metadata={}, created_at=datetime.datetime(2025, 7, 20, 18, 0, 21, 4900, tzinfo=datetime.timezone.utc), content='TERMINATE', type='TextMessage'), 
    
    TextMessage(id='10109946-dec3-4594-b6eb-5f9d030cdc94', source='CODE_EDITOR', models_usage=RequestUsage(prompt_tokens=210, completion_tokens=203), metadata={}, created_at=datetime.datetime(2025, 7, 20, 18, 0, 26, 164802, tzinfo=datetime.timezone.utc), content='```python\ndef binary_search(arr, target):\n    # Initialize left and right pointers\n    left, right = 0, len(arr) - 1\n    \n    # Loop until pointers converge\n    while left <= right:\n        # Calculate mid index\n        mid = left + (right - left) // 2\n        \n        # Check if target is found\n        if arr[mid] == target:\n            return mid  # Return index if found\n        \n        # Adjust search range\n        elif arr[mid] < target:\n            left = mid + 1  # Search right half\n        else:\n            right = mid - 1  # Search left half\n            \n    return -1  # Return -1 if not found\n\n# Example usage\narr = [1, 2, 3, 4, 5]  # Sorted array\ntarget = 3  # Element to search\nprint(binary_search(arr, target))  # Output: 2 (index of target)\n```', type='TextMessage')] 
    
    stop_reason='Maximum number of turns 3 reached.'